### PR TITLE
[Feature] Introduce Incremental Task Executor

### PR DESCRIPTION
**Pull request by LiShuMing**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #13880
- Make StartEpoch/StopMaintainece RPC work;
- Move StreamEpochManager to pipeline to support `report_epoch` in TaskExecutor.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 3 / 5 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|0|1|00.00%|[256]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|3|4|75.00%|[1487]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 3 / 5 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|0|1|00.00%|[256]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|3|4|75.00%|[1487]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 3 / 5 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|0|1|00.00%|[256]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|3|4|75.00%|[1487]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 3 / 5 (60.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|0|1|00.00%|[256]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|3|4|75.00%|[1487]|

**Comment by LiShuMing** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java [hunk 1]
```diff
@@ -1482,6 +1483,11 @@ private void toThriftForCommonParams(TExecPlanFragmentParams commonParams,
             } else {
                 commonParams.params.setNum_senders(instanceExecParams.size());
             }
+            if (isEnableStreamPipeline) {
+                commonParams.setIs_stream_pipeline(true);
+            } else {
+                commonParams.setIs_stream_pipeline(false);
+            }
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/qe/CoordinatorPreprocessor.java:1490]: commonParams.setIs_stream_pipeline(isEnableStreamPipeline);

File: be/src/exec/pipeline/exec_state_reporter.h [hunk 2]
```diff
@@ -29,14 +29,22 @@
...
                                                                     bool done);
     static Status report_exec_status(const TReportExecStatusParams& params, ExecEnv* exec_env,
                                      const TNetworkAddress& fe_addr);
-    ExecStateReporter();
+
     void submit(std::function<void()>&& report_task);
 
+    // STREAM MV
+    static TMVMaintenanceTasks create_report_epoch_params(QueryContext* query_ctx,
+                                                          std::vector<FragmentContext*> fragment_ctxs);
```
Comment (ABingHuang) [inline be/src/exec/pipeline/exec_state_reporter.h:44]:
```suggestion
                                                          const std::vector<FragmentContext*>& fragment_ctxs);
```

File: be/src/exec/pipeline/exec_state_reporter.h [hunk 3]
```diff
@@ -29,14 +29,22 @@
...
                                                                     FragmentContext* fragment_ctx, const Status& status,
                                                                     bool done);
     static Status report_exec_status(const TReportExecStatusParams& params, ExecEnv* exec_env,
                                      const TNetworkAddress& fe_addr);
-    ExecStateReporter();
+
     void submit(std::function<void()>&& report_task);
 
+    // STREAM MV
+    static TMVMaintenanceTasks create_report_epoch_params(QueryContext* query_ctx,
```
Comment (ABingHuang) [inline be/src/exec/pipeline/exec_state_reporter.h:43]:
```suggestion
    static TMVMaintenanceTasks create_report_epoch_params(const QueryContext* query_ctx,
```

File: be/src/exec/pipeline/pipeline_driver_poller.cpp [hunk 4]
```diff
@@ -192,6 +192,19 @@ size_t PipelineDriverPoller::activate_parked_driver(const ImmutableDriverPredica
...
 }
 
+size_t PipelineDriverPoller::calculate_parked_driver(const ImmutableDriverPredicateFunc& predicate_func) const {
+    size_t parked_driver_num = 0;
+    auto driver_it = _parked_drivers.begin();
+    while (driver_it != _parked_drivers.end()) {
+        auto driver = *driver_it;
+        if (predicate_func(driver)) {
+            parked_driver_num += 1;
+            driver_it++;
```
Comment (ABingHuang) [inline be/src/exec/pipeline/pipeline_driver_poller.cpp:202]: why ++ here? dead loop?

File: be/src/exec/pipeline/stream_epoch_manager.cpp [hunk 5]
```diff
@@ -96,4 +133,14 @@ const BinlogOffset* StreamEpochManager::_get_epoch_unlock(const TabletId2BinlogO
     }
 }
 
-} // namespace starrocks
+size_t StreamEpochManager::activate_parked_driver(ExecEnv* exec_env, const TUniqueId& query_id) {
+    // TODO(lism): how to decide use driver_executor or wg_driver_executor?
+    size_t num_activated_drivers = 0;
+    num_activated_drivers += exec_env->driver_executor()->activate_parked_driver(
+            [query_id](const pipeline::PipelineDriver* driver) { return driver->query_ctx()->query_id() == query_id; });
+    num_activated_drivers += exec_env->wg_driver_executor()->activate_parked_driver(
```
Comment (ABingHuang) [inline be/src/exec/pipeline/stream_epoch_manager.cpp:141]: for one query, it must be executed by driver_executor() or wg_driver_executor(). why should active drivers of both here?

**Review by ABingHuang — CHANGES_REQUESTED** [review]

File: be/src/exec/pipeline/fragment_context.cpp [hunk 6]
```diff
@@ -261,21 +261,19 @@ void FragmentContext::destroy_pass_through_chunk_buffer() {
...
     return Status::OK();
 }
 
 void FragmentContext::count_down_epoch_pipeline(RuntimeState* state, size_t val) {
-    VLOG_ROW << "count_down_epoch_pipeline"
-             << ", num_finished_epoch_pipelines:" << _num_finished_epoch_pipelines
-             << ", pipeline size:" << _pipelines.size();
-
-    bool all_pipelines_finished = _num_finished_epoch_pipelines.fetch_add(val) + val == _pipelines.size();
+    bool all_pipelines_finished = _num_finished_epoch_pipelines.fetch_add(val) + 1 == _pipelines.size();
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/fragment_context.cpp:271]: why not the ` + val` ?

File: be/src/exec/pipeline/pipeline_driver_poller.cpp (hunk 4 above)
Comment (mofeiatwork) [reply be/src/exec/pipeline/pipeline_driver_poller.cpp:202]: I think it's used to iterate all parked drivers, in the list

**Review by murphyatwork — DISMISSED** [review]

File: be/src/exec/pipeline/pipeline_driver_poller.cpp (hunk 4 above)
Comment (LiShuMing) [reply be/src/exec/pipeline/pipeline_driver_poller.cpp]: fixed.

File: be/src/exec/pipeline/stream_epoch_manager.cpp (hunk 5 above)
Comment (LiShuMing) [reply be/src/exec/pipeline/stream_epoch_manager.cpp]: fixed.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 3 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|0|1|00.00%|[256]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|2|2|100.00%|[]|

File: be/src/exec/pipeline/fragment_context.cpp (hunk 6 above)
Comment (LiShuMing) [reply be/src/exec/pipeline/fragment_context.cpp]: fixed.

File: be/src/exec/pipeline/stream_epoch_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        if (!_fragment_id_to_node_id_scan_ranges.empty()) {
+            if (_fragment_id_to_node_id_scan_ranges.size() != fragment_id_to_node_id_scan_ranges.size()) {
+                return Status::InternalError(fmt::format("Update epoch failed: {} is not equal to origin's size {}",
+                                                         fragment_id_to_node_id_scan_ranges.size(),
+                                                         _fragment_id_to_node_id_scan_ranges.size()));
+            }
+            for (auto& [fragment_instance_id, node_id_to_scan_ranges] : fragment_id_to_node_id_scan_ranges) {
+                auto iter = _fragment_id_to_node_id_scan_ranges.find(fragment_instance_id);
+                if (iter == _fragment_id_to_node_id_scan_ranges.end()) {
+                    return Status::InternalError(
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply be/src/exec/pipeline/stream_epoch_manager.cpp:40]: This is controlled by FE's Coordinator, should retry the pr by its StateMachine.

File: be/src/exec/pipeline/stream_epoch_manager.cpp [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+            }
+        } else {
+            _fragment_id_to_node_id_scan_ranges = fragment_id_to_node_id_scan_ranges;
+        }
+
+        _epoch_info = epoch_info;
+        _finished_fragment_ctxs.clear();
+
+        size_t num_drivers = 0;
+        bool enable_resource_group = true;
```
[the remark this answers is not in the document]
Comment (LiShuMing) [reply be/src/exec/pipeline/stream_epoch_manager.cpp]: yep, however `enable_resource_group ` is a  FragmentInstance param level config for now.  

I add a TODO that maybe later we can add a query level `enable_resource_group ` config later.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 3 (66.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/mv/MVMaintenanceJob.java|0|1|00.00%|[256]|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|2|2|100.00%|[]|

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Review by murphyatwork — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|2|2|100.00%|[]|

File: be/src/service/internal_service.cpp [hunk 9]
```diff
@@ -757,6 +766,28 @@ Status PInternalServiceImplBase<T>::_mv_start_maintenance(const TMVMaintenanceTa
         RETURN_IF_ERROR(fragment_executor.execute(_exec_env));
     }
 
+    // Prepare EpochManager
+    const TUniqueId& query_id = task.query_id;
+    auto&& existing_query_ctx = _exec_env->query_context_mgr()->get(query_id);
+    if (!existing_query_ctx) {
```
Comment (mofeiatwork) [inline be/src/service/internal_service.cpp:772]: Should it be `if (existing_query_ctx)` ?

Because it should not be existed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/CoordinatorPreprocessor.java|2|2|100.00%|[]|

**Review by murphyatwork — APPROVED** [review]

**Comment by LiShuMing** [comment]
run starrocks_be_unittest

**Comment by LiShuMing** [comment]
run starrocks_be_unittest

**Comment by LiShuMing** [comment]
run starrocks_be_unittest

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**